### PR TITLE
direct response support environment_variable

### PR DESCRIPTION
**Pull request by StarryVae**
Commit Message: direct response support environment_variable
Additional Description: N/A
Risk Level: low
Testing: ut
Docs Changes: N/A
Release Notes: N/A

### CODE REVIEW

**Comment by StarryVae** [comment]
issue #29219

File: test/common/router/config_impl_test.cc
Comment (adisuissa) [inline test/common/router/config_impl_test.cc:7581]: Consider using an intentional env-var over the well know PATH, e.g.:
```
  TestEnvironment::setEnvVar("ResponseBodyContents", "some body", 1);
  Envoy::Cleanup cleanup([]() { TestEnvironment::unsetEnvVar("ResponseBodyContents"); });
```

**Review by adisuissa — COMMENTED** [review]
Thanks for fixing that!
LGTM, modulo small nit on the tests, and please add a release note.

Assigning a senior-maintainer due to router changes.
/assign-from @envoyproxy/senior-maintainers

File: test/common/router/config_impl_test.cc
Comment (StarryVae) [reply test/common/router/config_impl_test.cc]: done, thanks.

**Comment by StarryVae** [comment]
friendly ping @adisuissa @lizan

**Review by adisuissa — DISMISSED** [review]
LGTM, thanks!

Assigning Ryan for a senior-maintainer pass.
/assign @RyanTheOptimist 
/unassign @lizan

File: source/common/router/config_utility.cc
Comment (RyanTheOptimist) [inline source/common/router/config_utility.cc:113]: Wow, this is much cleaner. Awesome!

I'm slightly concerned, though, about the potential for Envoy to run out of memory with this change. With the old code, in the case of a file we check the size before we allocate the string which contains it. With this new code, we'll read entire file into memory. This seems a bit risky. Is there any way we could get similar behavior? Maybe we should change DataSource::read() to take a max_size argument? I think this would be a fairly straightforward change and would avoid the OOM potential here.

Comment (StarryVae) [reply source/common/router/config_utility.cc:113]: yes, but does it mean that anywhere in Envoy using `DataSource::read()` has this risk? Is there a goog way to fix them all?

Comment (StarryVae) [reply source/common/router/config_utility.cc:113]: cc @RyanTheOptimist

**Comment by StarryVae** [comment]
friendly ping @RyanTheOptimist

**Comment by RyanTheOptimist** [comment]
Sorry, /assign appears to not have actually assigned me so when i checked my assigned PRs, this did not show up.

/assign @RyanTheOptimist

**Comment by RyanTheOptimist** [comment]
Looks like this need a main merge. Can you resolve the conflicts?

File: source/common/router/config_utility.cc
Comment (RyanTheOptimist) [reply source/common/router/config_utility.cc:113]: > yes, but does it mean that anywhere in Envoy using DataSource::read() has this risk? Is there a goog way to fix them all?

Yes, this is potentially an issue, though presumably only for data sources which are file-backed (and hence where the input is potentially enormous). I'm not sure it needs to block this PR, but it does seem worth filing an issue about.

Comment (StarryVae) [reply source/common/router/config_utility.cc:113]: yeah, i think so, i will submit a new issue about the risk, and it will not be fixed in this pr, thanks.

**Comment by StarryVae** [comment]
done, PTAL gain, thanks. @RyanTheOptimist

File: source/common/router/config_utility.cc
Comment (RyanTheOptimist) [reply source/common/router/config_utility.cc:116]: Well, I think we need to fix this *new* issue that we added in the PR since it changes behavior in a risky way. We can fix the *existing* issues in a followup.

Comment (StarryVae) [reply source/common/router/config_utility.cc:116]: ok, so maybe we should not change `DataSource::read()` to take a max_size argument here? since it will affect anywhere using it. And what about keeping the file process code here, and then add `DataSource::read()` to support `environment_variable` and `inline string`?

Comment (RyanTheOptimist) [reply source/common/router/config_utility.cc:133]: If we don't want to change all the callers of DataSource::read() now, then we could add the new max_size parameter with a default value of 0 (or -1, or some such). That seems like a pretty clean solution.

Comment (StarryVae) [reply source/common/router/config_utility.cc:133]: yeah, i think it is a greater solution, i will fix it.

**Review by RyanTheOptimist — COMMENTED** [review]
Looking great. One small suggestion.

File: source/common/config/datasource.h
Comment (StarryVae) [reply source/common/config/datasource.h:33]: done, PTAL again, thanks.

**Review by RyanTheOptimist — APPROVED** [review]

**Pull request merged**
